### PR TITLE
lodash use typescript Pick<> to improve _.pick typings

### PR DESCRIPTION
**Pull request by andnp**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

This provides the necessary overloads to provide correct and complete typings for `_.pick` by annotation or by type inference.

```
const o = {
  key1: [1, 2, 3],
  key2: "hello world"
}
let x = _.pick(o, ["key1"]);
x = _.pick(o, "key1");

// typeof x === {key1: number[]}

const key = "key1";
let y = _.pick(o, key);
y = _.pick(o, [key]);

// typeof y === {key1: number[]}

const keys: string[] = ["key1", "key2"];
let z = _.pick(o, keys);

/*
typeof z === Partial<{
  key1: number[],
  key2: string
}>
*/
```

### CODE REVIEW

**Comment by thorn0** [comment]
Mapped types can't be used for a number of lodash's functions including `pick` because those functions support deep property paths:
```js
var o = { a: { b: 1, c: 2 }, d: 3 };
_.pick(o, 'a.b'); // { a: { b: 1 } }
```

**Comment by andnp** [comment]
These typings should still be robust in the face of deep property paths.
```
var o = { a: { b: 1, c: 2}, d: 3};
var y = _.pick(o, 'a.b');
/*
typeof y = Partial<{
  a: { b: number, c: number}, d: number
}>
*/
```

which I believe is the current behavior?

These overrides I believe will just improve the typings if the string literal happens to be a key of the object, and will resort to the previous typings otherwise.

File: types/lodash/index.d.ts
Comment (thorn0) [inline types/lodash/index.d.ts]: Why do we need this overload?

**Comment by aj-r** [comment]
@andnp Please add tests to `lodash-tests.ts` that demonstrate the use cases for your new overloads.

File: types/lodash/index.d.ts
Comment (aj-r) [reply types/lodash/index.d.ts]: `PorpertyPath` already has type `Many<PropertyName>`, which basically means `PropertyName | PropertyName[]`. So`PropertyName | PropertyPath[] = `PropertyName | (PropertyName | PropertyName[])[]`, which is not right.

I think this overload is correct, and the previous overload (`props: PropertyPath[]`) is not.

File: types/lodash/index.d.ts
Comment (aj-r) [reply types/lodash/index.d.ts]: `U` is not used except in the return value. This overload is basically the same as the previous one, except this one violates  the [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes). I think this overload should be removed.

File: types/lodash/index.d.ts
Comment (aj-r) [reply types/lodash/index.d.ts]: Sorry, I mis-read your comment the first time. You are correct - we should merge these overloads.

File: types/lodash/index.d.ts
Comment (aj-r) [inline types/lodash/index.d.ts]: Merge this overload with the previous one:

```
pick<T extends object, U extends keyof T>(
    object: T | null | undefined,
    ...props: Many<U>
): Pick<T, U>;
```

File: types/lodash/index.d.ts
Comment (andnp) [reply types/lodash/index.d.ts]: Good call on that. I had missed it in my initial testing, I'll merge these overloads.

File: types/lodash/index.d.ts
Comment (andnp) [reply types/lodash/index.d.ts]: Yup, we don't need this overload. I'll delete this one

**Comment by RyanCavanaugh** [comment]
This PR has been open and unchanged 5 days without signoff or complaint. This will be merged by a maintainer soon if there are no objections.

**Review by aj-r — APPROVED** [review]

**Comment by connor4312** [comment]
This broke stuff.

```ts
import { pick } from 'lodash';

interface IGreeter {
    [species: string]: (name: string) => string;
}

const greet = pick<IGreeter>(
    { human: (name: string) => `hello ${name}` },
    ['human']
);

console.log(greet.human('connor'));
```

Errors with:

```
error TS2349: Cannot invoke an expression whose type lacks a call signature. Type 'PartialDeep<(name: string) => string>' has no compatible call signatures.
```

File: types/lodash/index.d.ts
Comment (andy-ms) [inline types/lodash/index.d.ts]: How can this take `undefined` as input but return a defined output?

Comment (thorn0) [reply types/lodash/index.d.ts]: `_.pick(undefined)` returns an empty object.

Comment (andy-ms) [reply types/lodash/index.d.ts]: The type will still be wrong because `_.pick(undefined, ["x"])` will not have an `x` property. It would make sense to only accept `undefined` in the second overload that returns `Partial`.

**Pull request merged**
